docs: add Role-Based Access Control page; refresh permissions docs (#23133)
mlflow/mlflow: docs/docs/self-hosting/security/basic-http-auth.mdx

The default permission for all users is `READ`. It can be changed in the [configuration](#configuration) file.

Permissions are granted via **roles** — see [Role-Based Access Control](/self-hosting/security/role-based-access-control)
for the model, scenarios, and API reference. For workspace-wide access controls
specifically, see [Workspace Permissions](/self-hosting/workspaces/permissions).
Supported resources include `Experiment`, `Registered Model`, and `Scorer`.
To access an API endpoint, a user must have the required permission.
Otherwise, a `403 Forbidden` response will be returned.

:::warning Legacy permission-CRUD endpoints removed

The pre-RBAC per-resource permission endpoints
(`/mlflow/{experiments,registered-models,scorers,gateway/*}/permissions/*` and
`/mlflow/workspaces/<workspace>/permissions`) and the matching
`AuthServiceClient` methods (`create_experiment_permission()`,
`set_workspace_permission()`, etc.) have been removed. Manage permissions
through the [role API](/self-hosting/security/role-based-access-control#api-reference)
or the admin UI. The matrix tables below still reference these endpoints in

> some rows; treat those as historical and use the role API instead.

Author: Fixed in [fa2dd3499](https://github.com/mlflow/mlflow/pull/23133/commits/fa2dd3499). Updated the warning admonition in `basic-http-auth.mdx` and the migration table in `role-based-access-control.mdx` to distinguish **deprecated** (per-resource: `/mlflow/{experiments,registered-models,scorers,gateway/*}/permissions/*` plus `create_experiment_permission()` etc. — still work, emit a deprecation warning, backed by synthetic per-user role grants) from **removed** (workspace-permission endpoints + the four `*_workspace_permission` client methods — calls 404).

🤖 *Claude response*

The default permission for all users is `READ`. It can be changed in the [configuration](#configuration) file.

> Permissions are granted via **roles**. See [Role-Based Access Control](/self-hosting/security/role-based-access-control)

Reviewer: Let's explicitly mention from which MLflow version we support RBAC

Author: Added in `ae67927d4`: "the role-based model and the role API were introduced in MLflow 3.12; the admin UI documented on the RBAC page is available from the next minor release." If 3.12 is wrong let me know and I'll re-pin against the actual release tag.

🤖 *Claude response*

Author: Pinned in 7af69cdb6: "available from MLflow 3.13.0", and added "MLflow versions before 3.13 only support the legacy per-resource permission APIs". 🤖 *Claude response*

for the model, scenarios, and API reference. For workspace-wide access controls
specifically, see [Workspace Permissions](/self-hosting/workspaces/permissions).
Supported resources include `Experiment`, `Registered Model`, and `Scorer`.
To access an API endpoint, a user must have the required permission.
Otherwise, a `403 Forbidden` response will be returned.

:::warning[Pre-RBAC permission management is changing]

> The following are **deprecated**. They still work (backed internally by

Reviewer: q: For the role API, what's the concrete migration path? Let's say if I only know the username and experiment ID, how should I grant experiment access to the specific user going forward? I think this use case should be clearly documented.

Author: Added an inline migration recipe inside the deprecation warning in `ae67927d4`. Shows the `create_role` + `add_role_permission` + `assign_role` pattern for `(username, experiment_id)` and notes how to substitute `registered_model`, `scorer`, or a gateway resource type. Cross-links to "Migrating from the pre-RBAC model" for the full surface mapping.

🤖 *Claude response*

Reviewer: To clarify, I can think of several paths:

1. Keep using existing per-user APIs for managing per-user permissions for now, then remove the "deprecation" message here
2. Ask users to create a role and give permission to the role
3. Create new general APIs for per user permission management (/mlflow/users/permissions/add) that uses a synthetic role, and document this as a migration path

Author: Yes — exactly this use case is what `grant_user_permission` is designed for. The migration recipe in the admonition (7af69cdb6) is now a one-liner:

```python
auth_client.grant_user_permission(username, "experiment", experiment_id, "EDIT")
```

No role to author, no assign step — internally it writes to the user's `__user_<id>__` synthetic role, same storage path the deleted `create_experiment_permission` used. Gated by per-resource MANAGE on the target resource, so an experiment owner can still delegate without holding workspace-wide MANAGE.

The shipped implementation is mlflow/mlflow#23247.

🤖 *Claude response*

synthetic per-user role grants), but emit a deprecation warning. Prefer the
[role API](/self-hosting/security/role-based-access-control#api-reference)
or the admin UI for new code.

- REST endpoints:
  - `/mlflow/experiments/permissions/*`
  - `/mlflow/registered-models/permissions/*`
  - `/mlflow/scorers/permissions/*`
  - `/mlflow/gateway/secrets/permissions/*`
  - `/mlflow/gateway/endpoints/permissions/*`
  - `/mlflow/gateway/model-definitions/permissions/*`
- `AuthServiceClient` methods: `create_experiment_permission()`,
  `update_registered_model_permission()`, and the ~23 sibling per-resource
  methods.

The following have been **removed** (calls return 404):

- REST endpoints: `/mlflow/workspaces/<workspace>/permissions/*`.
- `AuthServiceClient` methods: `set_workspace_permission()`,
  `list_workspace_permissions()`, `delete_workspace_permission()`,
  `list_user_workspace_permissions()`.

Use the seeded `<workspace>/admin` and `<workspace>/user` roles, or a custom
role with a `(workspace, *, …)` permission. See
[Role-Based Access Control](/self-hosting/security/role-based-access-control)
and [Workspace Permissions](/self-hosting/workspaces/permissions).

> :::

Reviewer: Can we link to "Migrating from the pre-RBAC model" section?

Author: Added in `ae67927d4` — the bottom of the warning admonition now links to `#migrating-from-the-pre-rbac-model` on the RBAC page.

🤖 *Claude response*

Author: Done in 7af69cdb6 — the warning admonition now ends with a cross-link to the "Migrating from the legacy permission model" section on the RBAC page (and to Workspace Permissions). The section heading was also renamed from "Migrating from the pre-RBAC model" → "Migrating from the legacy permission model" per Yuki's feedback on `pre-RBAC` being jargon. 🤖 *Claude response*

> 

Reviewer 2: Shall we remove above sentence and just say `Permissions are granted via roles` with reference?

role API (when the same permission set is shared across users):

```python
# Pre-RBAC:
#   auth_client.create_experiment_permission(experiment_id, username, "EDIT")
# After RBAC:
auth_client.grant_user_permission(username, "experiment", experiment_id, "EDIT")
```

Substitute `registered_model`, `prompt`, `scorer`, or a gateway resource
type as needed.
`grant_user_permission` is gated by per-resource MANAGE on the target resource,
which is the same authority check the legacy methods used.
An experiment owner can continue to delegate without holding workspace-wide MANAGE.

For workspace-tier grants (previously
`set_workspace_permission(workspace, user, permission)`), assign the
seeded `<workspace>/admin` or `<workspace>/user` role, or build a custom
role with a `(workspace, *, …)` permission and assign it. See
[Role-Based Access Control](/self-hosting/security/role-based-access-control),
[Migrating from the legacy permission model](/self-hosting/security/role-based-access-control#migrating-from-the-legacy-permission-model),
and [Workspace Permissions](/self-hosting/workspaces/permissions) for
the full surface mapping.

> 

Reviewer 2: ```suggestion
See [Migration from the legacy permission model](...) for the detailed guidance on how to migrate to the new RBAC permission model.
```

for simplification

The default permission for all users is `READ`. It can be changed in the [configuration](#configuration) file.

Permissions are granted via **roles**. See [Role-Based Access Control](/self-hosting/security/role-based-access-control)
for the model, scenarios, and API reference. For workspace-wide access controls
specifically, see [Workspace Permissions](/self-hosting/workspaces/permissions).
Supported resources include `Experiment`, `Registered Model`, and `Scorer`.
To access an API endpoint, a user must have the required permission.
Otherwise, a `403 Forbidden` response will be returned.

Note: the role-based model and the role API were introduced in MLflow 3.12;
the admin UI documented on the RBAC page is available from the next minor
release. Older MLflow versions only support the pre-RBAC per-resource
permission APIs documented below.

:::warning[Pre-RBAC permission management is changing]

The following are **deprecated**. They still work (backed internally by

> synthetic per-user role grants), but emit a deprecation warning. Prefer the

Reviewer 2: ```suggestion
synthetic per-user role grants), but may be removed in the future release. Prefer the
```

"emit a warning" is not very informative

Author: The whole admonition is now rewritten in 7af69cdb6 — the deprecation-warning framing is gone (decision locked with mprahl on rfcs#16, M6 + M9: drop the deprecation window entirely; the legacy methods are removed at the upgrade). The new admonition heading is "Pre-RBAC permission management has been removed" and the body describes the one-line replacement (`grant_user_permission`) plus the auth-store backfill migration that preserves existing grants. 🤖 *Claude response*

permission, and assign the user:

```python
role = auth_client.create_role(
    name=f"{username}-exp-{experiment_id}",
    workspace="default",
)
auth_client.add_role_permission(
    role_id=role.id,
    resource_type="experiment",
    resource_pattern=str(experiment_id),
    permission=permission,
)
auth_client.assign_role(username=username, role_id=role.id)
```

Substitute `registered_model`, `scorer`, or a gateway resource type for
other resources, and `workspace=` accordingly when workspaces are enabled.
See [Migrating from the pre-RBAC model](/self-hosting/security/role-based-access-control#migrating-from-the-pre-rbac-model)
for the full surface mapping.

The following have been **removed** (calls return 404):

- REST endpoints: `/mlflow/workspaces/<workspace>/permissions/*`.
- `AuthServiceClient` methods: `set_workspace_permission()`,
  `list_workspace_permissions()`, `delete_workspace_permission()`,
  `list_user_workspace_permissions()`.

Use the seeded `<workspace>/admin` and `<workspace>/user` roles, or a custom
role with a `(workspace, *, …)` permission. See
[Role-Based Access Control](/self-hosting/security/role-based-access-control),

> [Migrating from the pre-RBAC model](/self-hosting/security/role-based-access-control#migrating-from-the-pre-rbac-model),

Reviewer 2: nit: It might be only me, but `pre-RBAC` sounds a bit jargon, shall we rephrase it to sth like `legacy permission model before RBAC support`?

Author: Good point. Renamed `pre-RBAC` → `legacy permission model` / `legacy` across the docs PR (7af69cdb6) and the RFC (rfcs#16, 0df1c50) — 10 occurrences total. The migration section heading is now "Migrating from the legacy permission model" and the body pins "from a release before MLflow 3.13". Anchor changes from `#migrating-from-the-pre-rbac-model` to `#migrating-from-the-legacy-permission-model`; the link in the warning admonition is updated to match. 🤖 *Claude response*

MLflow provides <Link to="/api_reference/auth/rest-api.html#create-user" target="_blank">REST APIs</Link>
and a client class <APILink fn="mlflow.server.auth.client.AuthServiceClient">`AuthServiceClient`</APILink> to manage users and permissions.
Permissions are granted via roles; see [Role-Based Access Control](/self-hosting/security/role-based-access-control)
for the full model and the per-scenario examples (one-off direct grants, team
roles, workspace admins, onboarding). For workspace-scoped permissions, see

> <Link to="/self-hosting/workspaces/permissions">Workspace Permissions</Link>.

Author: Fixed in [1cc08c68d](https://github.com/mlflow/mlflow/pull/23133/commits/1cc08c68d). Tried removing the blank line first, but prettier (correctly, per MDX) reinserts it so the `<Link>` parses as JSX. Switched the inline reference to plain markdown — `[Workspace Permissions](/self-hosting/workspaces/permissions)` — which keeps the link inline within the paragraph and passes `yarn prettier:check`.

🤖 *Claude response*
